event/broker.go

```diff
@@ -0,0 +1,127 @@
+package event
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// Broker provides a friendly interface to fan-out events to as many subscribers
+// registered with it. Notice that an explicit unsubscription is necessary even
+// with a context.Context being sent on the subscription call (to be improved).
+type Broker interface {
+	Subscribe(ctx context.Context) (EventSource, error)
+	Unsubscribe(src EventSource) (hasMore bool, err error)
+}
+
+type broker struct {
+	ctx    context.Context
+	source EventSource
+
+	lock    sync.Mutex
+	clients []client
```

Comment:
Client order is not important here, what about using a `map[client]bool` to simplify (un)subscription?

Reply:
I think the reason for the array is that the expectation to iterate over the clients list much more than we actually sub/unsubscribe clients to them. So then the especially worse unsubscription performance compensates for the multiple faster iterations when delivering the events (I think iterating over a map is >10x slower)